Automigrate: Prefer framework import

ndelangen opened the pull request:
## What I did

This pull request to `code/lib/cli-storybook` introduces new functionality to automate the migration from renderer packages to framework packages. The changes include adding new helper functions, updating existing functions, and adding comprehensive tests.

Key changes include:

### New Functionality:
* [`code/lib/cli-storybook/src/automigrate/fixes/renderer-to-framework.ts`](diffhunk://#diff-fdbb7e210404b175efe6b5371fda16dd381b07814d9e4ae9c8823b49455aa095R1-R244): Added functions to detect frameworks and renderers, transform source files, remove renderer packages from `package.json`, and prompt the user for framework selection.

### Testing:
* [`code/lib/cli-storybook/src/automigrate/fixes/renderer-to-framework.test.ts`](diffhunk://#diff-f211b2f82cb76842581ba7b5a663d5f83b94cf704fe9604f2c6e2746af45ce70R1-R255): Added tests to verify the new functionality, including transforming source files, removing renderer packages from `package.json`, and handling various edge cases such as file read errors.

## Checklist for Contributors

### Testing

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to communicate how to test your changes -->

#### The changes in this PR are covered in the following automated tests:

- [ ] stories
- [x] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

<!-- Please include the steps to test your changes here. For example:

1. Run a sandbox for template, e.g. `yarn task --task sandbox --start-from auto --template react-vite/default-ts`
2. Open Storybook in your browser
3. Access X story

-->

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli-storybook/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

  - `bug`: Internal changes that fixes incorrect behavior.
  - `maintenance`: User-facing maintenance tasks.
  - `dependencies`: Upgrading (sometimes downgrading) dependencies.
  - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
  - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
  - `documentation`: Documentation **only** changes. Will not show up in release changelog.
  - `feature request`: Introducing a new feature.
  - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
  - `other`: Changes that don't fit in the above categories.

   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

<!-- CANARY_RELEASE_SECTION -->

<!-- BENCHMARK_SECTION -->
<!-- BENCHMARK_SECTION -->


<!-- greptile_comment -->

## Greptile Summary

Introduced new automated migration functionality to help users transition from Storybook renderer packages to framework packages, with comprehensive testing coverage.

- Added `renderer-to-framework.ts` with core migration logic for detecting frameworks/renderers and transforming imports
- Added parallel processing with `p-limit` for efficient source file transformation
- Added interactive framework selection prompt when multiple frameworks are detected
- Added comprehensive test suite covering file transformations, package.json updates, and error handling
- Added support for custom glob patterns to control which files are scanned



<!-- /greptile_comment -->

ndelangen committed:
Automigrate: Add renderer to framework migration fix

Introduces a new automigrate fix that helps users transition from renderer-specific packages to framework packages in Storybook 9.0. The fix:
- Detects renderer and framework packages in project
- Allows interactive selection of framework
- Transforms import statements
- Removes renderer dependencies from package.json

ndelangen committed:
Automigrate: Add tests for renderer to framework migration

Adds comprehensive test coverage for the renderer to framework migration fix, including:
- Source file transformation tests
- Package.json dependency removal tests
- Framework and renderer detection tests
- Error handling scenarios

ndelangen committed:
Merge branch 'next' into norbert/migration-to-framework-package-import

ndelangen committed:
Merge branch 'next' into norbert/migration-to-framework-package-import

ndelangen committed:
Merge branch 'next' into norbert/migration-to-framework-package-import

ndelangen commented on code/lib/cli-storybook/src/automigrate/fixes/renderer-to-framework.ts:
```diff
@@ -0,0 +1,244 @@
+      }
+    });
+
+    if (!dryRun && hasChanges) {
+      await writeFile(packageJsonPath, JSON.stringify(packageJson, null, 2));
+    }
```
comments in `package.json` files?
you wish!

valentinpalkovic commented on code/lib/cli-storybook/src/automigrate/fixes/renderer-to-framework.ts:
```diff
@@ -0,0 +1,244 @@
+
+  prompt(result: MigrationResult): string {
+    if (result.frameworks.length > 1) {
+      return dedent`
+        Multiple frameworks detected. You will be prompted to select which framework to use.
+        Would you like to proceed with the migration?
+      `;
```
It's not entirely clear for the user what's happening here. Let's add some more context. Also for the `else` branch

valentinpalkovic commented on code/lib/cli-storybook/src/automigrate/fixes/renderer-to-framework.ts:
```diff
@@ -0,0 +1,244 @@
+import { readFile, writeFile } from 'node:fs/promises';
```
Don't forget it to add it to the index file. Also please create a canary and test it once in a project of your choice before merging.

ndelangen committed:
Add renderer-to-framework migration fix

Introduced a new migration fix to transition from renderer-based to framework-based configuration in Storybook. Updated prompts to inform users about detected frameworks and the changes that will be made to their project, including updating imports, removing renderer packages, and installing necessary dependencies.
